File: apps/desktop/src/lib/types/session.ts
// Hand-mirrored from `capture-types/src/session.rs` (no codegen — the serde
// wire-pin test there is what keeps these spellings honest).
//
// `assumed_working` / `possibly_blocked` are system audio's, and only system
// audio's: a Core Audio process tap has no authorization query, so its state is
// inferred from whether a tap has ever delivered sound, never read (ADR 0052).
export type PermissionStatus =
	| "granted"
	| "denied"
	| "not_determined"
	| "restricted"
	| "unsupported"
	| "unknown"
	| "assumed_working"
	| "possibly_blocked";

export interface SupportedSources {
	screen: boolean;
	microphone: boolean;
	systemAudio: boolean;
}

export interface CaptureSupport {
	platform: string;
	nativeCaptureSupported: boolean;
	supportedSources: SupportedSources;
}

export interface PermissionsMap {
	screen: PermissionStatus;
	microphone: PermissionStatus;
	systemAudio: PermissionStatus;
}

export interface RequestedSources {
	screen: boolean;
	microphone: boolean;
	systemAudio: boolean;
}

export interface CaptureOutputFiles {
	screenFile: string | null;
	screenFiles: string[];
	microphoneFile: string | null;
	microphoneFiles: string[];
	systemAudioFile: string | null;
	systemAudioFiles: string[];
}

export interface SourceSessionMeta {
	sessionId: string;
	startedAtUnixMs: number;
}

export interface SourceSessions {
	screen: SourceSessionMeta | null;
	microphone: SourceSessionMeta | null;
	systemAudio: SourceSessionMeta | null;
}

export interface CaptureSession {
	isRunning: boolean;
	requestedSources: RequestedSources | null;
	/**
	 * Sources the user turned off mid-session — a user-scoped pause on one
	 * source, always a subset of `requestedSources`. What is actually recording
	 * is `requestedSources` minus this.
	 */
	maskedSources: RequestedSources;
	outputFiles: CaptureOutputFiles | null;
	sourceSessions: SourceSessions | null;
	isInactivityPaused: boolean;
	isUserPaused: boolean;
	isLowDiskSuspended: boolean;
}

export interface GetPermissionsResponse {
	permissions: PermissionsMap;
	session: CaptureSession | null;
}

export interface GeckoBrowserInfo {
	bundleId: string;
	displayName: string;
	installed: boolean;
}

export interface BrowserUrlAccessibilityStatus {
	trusted: boolean;
	geckoBrowsers: GeckoBrowserInfo[];
}

export interface StartCaptureResponse {
	session: CaptureSession;
}

export interface StopCaptureResponse {
	session: CaptureSession;
}

File: crates/capture-types/src/session.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CapturePermissionState {
    Granted,
    Denied,
    NotDetermined,
    Unsupported,
    Unknown,
    /// System audio only (ADR 0052). Core Audio process taps have their own TCC
    /// category and **no authorization query at all**, so the two states below
    /// are inferred from what the tap delivered, never read from the OS.
    ///
    /// A tap has delivered sound, which only a granted tap can do.
    AssumedWorking,
    /// A tap has run and never delivered a sound. Denial looks exactly like a
    /// quiet Mac, so this is a suspicion, not a verdict — the surfaces that
    /// render it say "may be blocked" and let the user dismiss it.
    PossiblyBlocked,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CaptureSources {
    pub screen: bool,
    pub microphone: bool,
    pub system_audio: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureSupportResponse {
    pub platform: String,
    pub native_capture_supported: bool,
    pub supported_sources: CaptureSources,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CapturePermissions {
    pub screen: CapturePermissionState,
    pub microphone: CapturePermissionState,
    pub system_audio: CapturePermissionState,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureOutputFiles {
    pub screen_file: Option<String>,
    pub screen_files: Vec<String>,
    pub microphone_file: Option<String>,
    pub microphone_files: Vec<String>,
    pub system_audio_file: Option<String>,
    pub system_audio_files: Vec<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceSessionMeta {
    pub session_id: String,
    pub started_at_unix_ms: u64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceSessions {
    pub screen: Option<SourceSessionMeta>,
    pub microphone: Option<SourceSessionMeta>,
    pub system_audio: Option<SourceSessionMeta>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeCaptureSession {
    pub is_running: bool,
    pub is_inactivity_paused: bool,
    pub is_user_paused: bool,
    pub is_low_disk_suspended: bool,
    pub requested_sources: Option<CaptureSources>,
    pub output_files: Option<CaptureOutputFiles>,
    pub source_sessions: Option<SourceSessions>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CapturePermissionsResponse {
    pub permissions: CapturePermissions,
    pub session: NativeCaptureSession,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartNativeCaptureRequest {
    pub capture_screen: bool,
    pub capture_microphone: bool,
    pub capture_system_audio: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeCaptureSessionResponse {
    pub session: NativeCaptureSession,
}

// There is no codegen between these types and their hand-written TypeScript
// mirror, so the wire strings are pinned here: `PermissionStatus` in
// `apps/desktop/src/lib/types/session.ts` has to spell them the same way.
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn permission_states_round_trip_their_wire_strings() {
        for (state, wire) in [
            (CapturePermissionState::Granted, "granted"),
            (CapturePermissionState::Denied, "denied"),
            (CapturePermissionState::NotDetermined, "not_determined"),
            (CapturePermissionState::Unsupported, "unsupported"),
            (CapturePermissionState::Unknown, "unknown"),
            (CapturePermissionState::AssumedWorking, "assumed_working"),
            (CapturePermissionState::PossiblyBlocked, "possibly_blocked"),
        ] {
            assert_eq!(serde_json::to_value(state).unwrap(), wire);
            assert_eq!(
                serde_json::from_value::<CapturePermissionState>(wire.into()).unwrap(),
                state
            );
        }
    }

    #[test]
    fn permissions_serialize_the_inferred_system_audio_states() {
        let permissions = CapturePermissions {
            screen: CapturePermissionState::Granted,
            microphone: CapturePermissionState::Denied,
            system_audio: CapturePermissionState::PossiblyBlocked,
        };

        assert_eq!(
            serde_json::to_value(&permissions).unwrap(),
            serde_json::json!({
                "screen": "granted",
                "microphone": "denied",
                "systemAudio": "possibly_blocked",
            })
        );
    }
}
